pandas/core/indexes/multi.py

```diff
@@ -3058,26 +3060,44 @@ def _update_indexer(idxr, indexer=indexer):
                 # a collection of labels to include from this level (these
                 # are or'd)
                 indexers = None
+                if not need_sort:
+                    # Find out if the list_like label are sorted as the levels or not
+                    k_codes = np.array(
+                        [self.levels[i].get_loc(e) for e in k if e in self.levels[i]]
+                    )
+                    need_sort = not (k_codes[:-1] < k_codes[1:]).all()
                 for x in k:
                     try:
                         idxrs = _convert_to_indexer(
                             self._get_level_indexer(x, level=i, indexer=indexer)
                         )
-                        indexers = idxrs if indexers is None else indexers | idxrs
+                        if need_sort and indexer is not None:
+                            # We intersect with indexer to make idxrs
+                            # ordered as previously seen indexes
+                            idxrs = indexer.intersection(idxrs)
+
+                        indexers = (
+                            idxrs
+                            if indexers is None
+                            else indexers.union(idxrs, sort=False)
+                        )
                     except KeyError:
-
                         # ignore not founds
                         continue
 
                 if indexers is not None:
-                    indexer = _update_indexer(indexers, indexer=indexer)
+                    if need_sort:
```

Comment:
line 3092 devolves to exactly ``indexers & indexers`` which is == ``indexers``, so i think you can skip this entire section

Reply:
I rename the indexerS variable to indexers_union to make the difference between the indexers and indexer variable more evident.